File: felicia/examples/learn/message_communication/cc/node_creator.cc
#include "felicia/cc/master_proxy.h"
#include "felicia/examples/learn/message_communication/cc/node_create_flag.h"
#include "felicia/examples/learn/message_communication/cc/simple_publishing_node.h"
#include "felicia/examples/learn/message_communication/cc/simple_subscribing_node.h"

namespace felicia {

int RealMain(int argc, char* argv[]) {
  NodeCreateFlag delegate;
  FlagParser parser;
  parser.set_program_name(argv[0]);
  if (!parser.Parse(argc, argv, &delegate)) {
    return 1;
  }

  MasterProxy& master_proxy = MasterProxy::GetInstance();
  Status s = master_proxy.Start();
  CHECK(s.ok());

  NodeInfo node_info;
  node_info.set_name(delegate.name());

  if (delegate.is_publishing_node()) {
    master_proxy.RequestRegisterNode<SimplePublishingNode>(
        node_info, delegate.topic(), delegate.channel_type());
  } else {
    master_proxy.RequestRegisterNode<SimpleSubscribingNode>(node_info,
                                                            delegate.topic());
  }

  master_proxy.Run();

  return 0;
}

}  // namespace felicia

int main(int argc, char* argv[]) { return felicia::RealMain(argc, argv); }

File: felicia/examples/learn/message_communication/protobuf/python/node_creator.py
import sys

import felicia.examples.learn.message_communication.common.python.import_order_resolver # Should be before import felicia_py
import felicia_py as fel
from felicia.core.protobuf.master_data_pb2 import NodeInfo
import felicia_py.command_line_interface as cli

from felicia.examples.learn.message_communication.common.python.node_create_flag import NodeCreateFlag
from simple_publishing_node import SimplePublishingNode
from simple_subscribing_node import SimpleSubscribingNode


def main():
    fel.felicia_init()

    delegate = NodeCreateFlag()
    parser = cli.FlagParser()
    parser.set_program_name('node_creator')
    if not parser.parse(len(sys.argv), sys.argv, delegate):
        sys.exit(1)

    s = fel.MasterProxy.start()
    if not s.ok():
        print("{} {}.".format(cli.RED_ERROR, s), file=sys.stderr)
        sys.exit(1)

    node_info = NodeInfo()
    node_info.name = delegate.name_flag.value

    if delegate.is_publshing_node_flag.value:
        fel.MasterProxy.request_register_node(
            SimplePublishingNode,
            node_info,
            delegate.topic_flag.value,
            delegate.channel_type_flag.value)
    else:
        fel.MasterProxy.request_register_node(
            SimpleSubscribingNode,
            node_info,
            delegate.topic_flag.value)

    fel.MasterProxy.run()


if __name__ == '__main__':
    main()
